presto-main/src/main/java/com/facebook/presto/operator/DriverYieldSignal.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+public class DriverYieldSignal
+{
+    private long runningSequence;
+    private boolean running;
+    private final AtomicBoolean yield = new AtomicBoolean();
+
+    public synchronized void yieldAfter(long maxRunNanos, ScheduledExecutorService executor)
+    {
+        checkArgument(!running, "driver was running");
+        checkArgument(!shouldYield(), "yield while driver was not running");
+        running = true;
+
+        this.runningSequence++;
+        long expectedRunningSequence = this.runningSequence;
+        executor.schedule(() -> {
+            yield(expectedRunningSequence);
+        }, maxRunNanos, NANOSECONDS);
+    }
+
+    public synchronized void reset()
+    {
+        checkArgument(running, "driver was not running");
```

Comment:
use `checkState` since this is not an argument, and is only invalid because of the state of this instance